core/ir/aarch64/encode.c

```diff
@@ -196,6 +196,177 @@ const reg_id_t dr_reg_fixer[] = { REG_NULL,
 
     DR_REG_CNTVCT_EL0,
 };
+
+/* Maps real ISA registers to their corresponding virtual DR_ISA_REGDEPS register.
+ * Note that we map real sub-registers to their corresponding containing virtual register.
+ * Same size as dr_reg_fixer[], keep them synched.
+ */
+const reg_id_t d_r_reg_id_to_virtual[] = {
+    DR_REG_NULL, /* DR_REG_NULL */
+    DR_REG_NULL, /* DR_REG_NULL */
+
+#define VIRTUAL_XREGS
+    DR_REG_V0, DR_REG_V1, DR_REG_V2, DR_REG_V3, DR_REG_V4, DR_REG_V5, DR_REG_V6,     \
+    DR_REG_V7, DR_REG_V8, DR_REG_V9, DR_REG_V10, DR_REG_V11, DR_REG_V12, DR_REG_V13, \
+    DR_REG_V14, DR_REG_V15, DR_REG_V16, DR_REG_V17, DR_REG_V18, DR_REG_V19,          \
+    DR_REG_V20, DR_REG_V21, DR_REG_V22, DR_REG_V23, DR_REG_V24, DR_REG_V25,          \
+    DR_REG_V26, DR_REG_V27, DR_REG_V28, DR_REG_V29, DR_REG_V30, DR_REG_V31,          \
+    DR_REG_V32,
+
+    VIRTUAL_XREGS /* from DR_REG_X0 to DR_REG_XZR */
+    VIRTUAL_XREGS /* from DR_REG_W0 to DR_REG_WZR */
+#undef VIRTUAL_XREGS
+
+#define VIRTUAL_ZREGS
+    DR_REG_V33, DR_REG_V34, DR_REG_V35, DR_REG_V36, DR_REG_V37, DR_REG_V38, DR_REG_V39, \
+    DR_REG_V40, DR_REG_V41, DR_REG_V42, DR_REG_V43, DR_REG_V44, DR_REG_V45, DR_REG_V46, \
+    DR_REG_V47, DR_REG_V48, DR_REG_V49, DR_REG_V50, DR_REG_V51, DR_REG_V52,             \
+    DR_REG_V53, DR_REG_V54, DR_REG_V55, DR_REG_V56, DR_REG_V57, DR_REG_V58,             \
+    DR_REG_V59, DR_REG_V60, DR_REG_V61, DR_REG_V62, DR_REG_V63, DR_REG_V64,
+
+    VIRTUAL_ZREGS /* from DR_REG_Z0 to DR_REG_Z31 */
+    VIRTUAL_ZREGS /* from DR_REG_Q0 to DR_REG_Q31 */
+    VIRTUAL_ZREGS /* from DR_REG_D0 to DR_REG_D31 */
+    VIRTUAL_ZREGS /* from DR_REG_S0 to DR_REG_S31 */
+    VIRTUAL_ZREGS /* from DR_REG_H0 to DR_REG_H31 */
+    VIRTUAL_ZREGS /* from DR_REG_B0 to DR_REG_B31 */
```

Comment:
(Much much easier to read and understand.)

Reply:
Great! :)